Kernel/Arch/i386/CPU.cpp

```diff
@@ -854,6 +855,114 @@ FlatPtr read_cr4()
     return cr4;
 }
 
+void read_debug_registers_into(DebugRegisterState& debug_register_state)
+{
+    debug_register_state.dr0 = read_dr0();
+    debug_register_state.dr1 = read_dr1();
+    debug_register_state.dr2 = read_dr2();
+    debug_register_state.dr3 = read_dr3();
+    debug_register_state.dr6 = read_dr6();
+    debug_register_state.dr7 = read_dr7();
+}
+
+void write_debug_registers_from(DebugRegisterState& state)
+{
+    write_dr0(state.dr0);
+    write_dr1(state.dr1);
+    write_dr2(state.dr2);
+    write_dr3(state.dr3);
+    write_dr6(state.dr6);
+    write_dr7(state.dr7);
+}
+
+FlatPtr read_dr0()
+{
+    FlatPtr dr0;
+#if ARCH(I386)
+    asm("mov %%dr0, %%eax"
+        : "=a"(dr0));
+#else
+    asm("mov %%dr0, %%rax"
+        : "=a"(dr0));
+#endif
+    return dr0;
+}
+
+void write_dr0(FlatPtr value)
+{
+#if ARCH(I386)
+    asm volatile("mov %%eax, %%dr0" ::"a"(value));
+#else
+    asm volatile("mov %%rax, %%dr0" ::"a"(value));
+#endif
+}
+
+FlatPtr read_dr1()
+{
+    FlatPtr dr1;
+#if ARCH(I386)
+    asm("mov %%dr1, %%eax"
+        : "=a"(dr1));
+#else
+    asm("mov %%dr1, %%rax"
+        : "=a"(dr1));
+#endif
+    return dr1;
+}
+
+void write_dr1(FlatPtr value)
+{
+#if ARCH(I386)
+    asm volatile("mov %%eax, %%dr1" ::"a"(value));
+#else
+    asm volatile("mov %%rax, %%dr1" ::"a"(value));
+#endif
+}
+
+FlatPtr read_dr2()
+{
+    FlatPtr dr2;
+#if ARCH(I386)
+    asm("mov %%dr2, %%eax"
+        : "=a"(dr2));
+#else
+    asm("mov %%dr2, %%rax"
+        : "=a"(dr2));
+#endif
+    return dr2;
+}
+
+void write_dr2(FlatPtr value)
+{
+#if ARCH(I386)
+    asm volatile("mov %%eax, %%dr2" ::"a"(value));
+#else
+    asm volatile("mov %%rax, %%dr2" ::"a"(value));
+#endif
+}
+
+FlatPtr read_dr3()
+{
+    FlatPtr dr3;
+#if ARCH(I386)
+    asm("mov %%dr3, %%eax"
```

Comment:
Given these are all the same besides the debug register name, perhaps we should generate their definition via macros?